ext/rotoscope/stack.c

```diff
@@ -0,0 +1,109 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "stack.h"
+
+static void check_buffer(rs_stack_frame_t *contents)
+{
+  if (contents == NULL)
+  {
+    fprintf(stderr, "Not enough memory to allocate stack\n");
+    exit(1);
+  }
+}
+
+static void insert_root_node(rs_stack_t *stack)
+{
+  rs_tracepoint_t root_trace = (rs_tracepoint_t) {
+      .entity = "<ROOT>",
+      .event = UNKNOWN_STR,
+      .method_name = UNKNOWN_STR,
+      .method_level = UNKNOWN_STR,
+      .filepath = UNKNOWN_STR,
+      .lineno = 0
+  };
+
+  stack_push(stack, root_trace);
+}
+
+static void resize_buffer(rs_stack_t *stack)
+{
+  unsigned int newsize = stack->capacity * 2;
+  stack->contents = realloc(stack->contents, sizeof(rs_stack_frame_t) * newsize);
```

Comment:
Did you consider using ruby_xrealloc and ruby_xmalloc so that it can garbage collect if there isn't available space and raise if it still can't allocate memory after garbage collection?  That way you could remove `check_buffer`